### PR TITLE
Feature/static category heading

### LINKED ISSUE #1181: Allow static use of `exp:category_heading`
**Is your feature request related to a problem? Please describe.**
A lot of time I need to get a category metadata, but I can't use `exp:category_heading`, I need to write a custom query.

**Describe the solution you'd like**
As `exp:channel` allows us to override URL detection, the same could be valid for `exp:category_heading`.

**Describe alternatives you've considered**
Without this feature, I need to write a custom query.

**Teachability, Documentation, Adoption, Migration Strategy**
The documentation is pending on a [pull request](https://github.com/ExpressionEngine/ExpressionEngine-User-Guide/pull/199)

- [x] I am capable and would like to work on implementation of this feature if it is considered.

### PR DESCRIPTION
**Pull request by robsonsobral**
## Overview

Allow `exp:category_heading`  to be used statically.

## Nature of This Change

- [ ] 🐛 Fixes a bug
- [x] 🚀 Implements a new feature
- [ ] 🛁 Refactors existing code
- [ ] 💅 Fixes coding style
- [ ] ✅ Adds tests
- [ ] 👽 Adds new dependency
- [ ] 🔥 Removes unused files / code
- [ ] 🔒 Improves security 

## Is this backwards compatible?

- [x] Yes
- [ ] No

## Documentation

User Guide Pull Request: https://github.com/ExpressionEngine/ExpressionEngine-User-Guide/pull/199

closes #1181

### CODE REVIEW

File: system/ee/ExpressionEngine/Addons/channel/mod.channel.php [hunk 1]
```diff
@@ -3986,9 +3986,13 @@ public function category_heading()
 
         // Is the category being specified by name?
 
-        if ($qstring != '' and $this->reserved_cat_segment != '' and in_array($this->reserved_cat_segment, explode("/", $qstring)) and ee()->TMPL->fetch_param('channel')) {
-            $qstring = preg_replace("/(.*?)\/" . preg_quote($this->reserved_cat_segment) . "\//i", '', '/' . $qstring);
-
+        if (
+            (
+                ($this->reserved_cat_segment != '' and in_array($this->reserved_cat_segment, explode("/", $qstring)))
+                OR ee()->TMPL->fetch_param('category_url_name')
```
Comment (intoeetive) [inline system/ee/ExpressionEngine/Addons/channel/mod.channel.php]: Can we use `category_url_title` here instead? `category_url_name` is not used anywhere else in EE

File: system/ee/ExpressionEngine/Addons/channel/mod.channel.php [hunk 2]
```diff
@@ -4025,6 +4029,8 @@ public function category_heading()
             }
 
             if ($valid == 'y') {
+                $qstring = preg_replace("/(.*?)\/" . preg_quote($this->reserved_cat_segment) . "\//i", '', '/' . $qstring);
```
Comment (intoeetive) [inline system/ee/ExpressionEngine/Addons/channel/mod.channel.php]: Is there some particular reason why this line of code has been moved down here?

**Review by intoeetive — CHANGES_REQUESTED** [review]
@robsonsobral Thank you for this PR!

I have added a couple of comments - would you mind taking care of those?

Also, is there a related issue on GitHub? If not, would you mind creating one? (so that we know what feature this adds)

File: system/ee/ExpressionEngine/Addons/channel/mod.channel.php (hunk 1 above)
Comment (robsonsobral) [reply system/ee/ExpressionEngine/Addons/channel/mod.channel.php]: Sure! I thought I already did it! I'm sorry!

File: system/ee/ExpressionEngine/Addons/channel/mod.channel.php (hunk 2 above)
Comment (robsonsobral) [reply system/ee/ExpressionEngine/Addons/channel/mod.channel.php]: Just because the `$qstring` isn't use anywhere until this block. I can move it back.

Comment (intoeetive) [reply system/ee/ExpressionEngine/Addons/channel/mod.channel.php]: It might be just safer to keep it where it was - so can you revert?

Comment (robsonsobral) [reply system/ee/ExpressionEngine/Addons/channel/mod.channel.php]: Sure!

**Review by intoeetive — COMMENTED** [review]
This looks good to me.

I'm going to test it (and likely write automated Cypress test - unless you're willing to do that) and then approve

**Comment by robsonsobral** [comment]
Please do, so I can read it and learn how to do it for the next time. Thank you.

**Review by intoeetive — CHANGES_REQUESTED** [review]
@robsonsobral I have added automated tests and unfortunately those fail ;(
It looks like it now wants to have BOTH category indicator and parameter present.
Please see the test templates and scenarios I've added

**Comment by robsonsobral** [comment]
@intoeetive , I'm sorry! It looks like the test dependencies don't like my work environment. I couldn't make them to work. Some Cypress issue... So, I tested it manually.

Do you mind to test it again?

**Comment by intoeetive** [comment]
@robsonsobral unfortunately that still does not work.

You can take the test templates I've pushed to Cypress folder and test using it in different scenarios: with category segment being or not being present, matching and not matching the passed parameter etc.

On the Cypress front, I'm interested to know what issues you have? I see many people have trouble using that, so I'll probably have to make some sort of tutorial or presentation on using it, but I would like to know what the common pitfalls are. There's some infomation about the setup needed here: https://github.com/ExpressionEngine/ExpressionEngine/blob/6.dev/CONTRIBUTING.md#tests

**Review by litzinger — COMMENTED** [review]
Should this be added to EE's ignore list? Seems to me like something that should be added to the user's global ignore list and not managed by EE itself.

**Comment by robsonsobral** [comment]
I'm sorry it took so long, @intoeetive .

I tested manually again, because NodeSASS don't like my computer. Maybe, after eecms move to DartSASS, I can make the tests to work.

**Comment by intoeetive** [comment]
@robsonsobral unfortunately there are still tests that are failing: 
image

What is the issue that you have with node? You might want to try some other version (I am running `10.22.1`). Also, SASS is not needed to run the tests

**Comment by robsonsobral** [comment]
I'm sorry, @intoeetive ! Please, forgive me for give you so much trouble. I tried a lot today, but when I try to install the tests, I get a Cypress error; and, when I try docker, I get a node-sass error. But I didn't give up yet! I'll not submit any commit without automatic tests, but I don't know how much time it will take!

Thank you for help and patience!

**Comment by intoeetive** [comment]
So what's the Cypress error that you get? Maybe I could help with that. Ping me in Slack if you see me online, or just post it here

**Comment by robsonsobral** [comment]
Does your test suite passed, @intoeetive ?

File: system/ee/ExpressionEngine/Addons/channel/mod.channel.php [hunk 3]
```diff
@@ -4056,15 +4072,19 @@ public function category_heading()
 
         // Is the category being specified by ID?
 
-        if (! preg_match("#(^|\/)C(\d+)#", $qstring, $match)) {
+        if (! preg_match("#(^|\/)C(\d+)#", $qstring, $match) AND ! ee()->TMPL->fetch_param('category_id'))
+        {
```
Comment (intoeetive) [inline system/ee/ExpressionEngine/Addons/channel/mod.channel.php]: According to PSR-12, opening brace should be on same line with `if`

File: system/ee/ExpressionEngine/Addons/channel/mod.channel.php [hunk 4]
```diff
@@ -4079,9 +4099,10 @@ public function category_heading()
         $query = ee()->db->query("SELECT c.cat_name, c.parent_id, c.cat_url_title, c.cat_description, c.cat_image {$field_sqla}
 							FROM exp_categories AS c
 							{$field_sqlb}
-							WHERE c.cat_id = '" . ee()->db->escape_str($match[2]) . "'");
+							WHERE c.cat_id = '" . ee()->db->escape_str($cat_id) . "'");
 
-        if ($query->num_rows() == 0) {
+        if ($query->num_rows() == 0)
+        {
```
Comment (intoeetive) [inline system/ee/ExpressionEngine/Addons/channel/mod.channel.php]: same as above

File: system/ee/ExpressionEngine/Addons/channel/mod.channel.php [hunk 5]
```diff
@@ -3951,7 +3951,7 @@ public function close_ul($parent_id, $depth = 0)
       */
     public function category_heading()
     {
-        if ($this->query_string == '') {
+        if ($this->query_string === '' && !ee()->TMPL->fetch_param('category_url_title') && !ee()->TMPL->fetch_param('category_id')) {
```
Comment (intoeetive) [inline system/ee/ExpressionEngine/Addons/channel/mod.channel.php]: Just out of curiosity, has there been specific reason to also check for type (`===`) of `$this->query_string`. I think it's always string, so should be safe to do, so just wondering

**Review by intoeetive — CHANGES_REQUESTED** [review]
With the exception of some PSR formatting that needs to be improved, I think we're good to go

File: system/ee/ExpressionEngine/Addons/channel/mod.channel.php (hunk 3 above)
Comment (robsonsobral) [reply system/ee/ExpressionEngine/Addons/channel/mod.channel.php]: Jesus... I was pretty sure it was the other way! Fixing it right now!

File: system/ee/ExpressionEngine/Addons/channel/mod.channel.php (hunk 5 above)
Comment (robsonsobral) [reply system/ee/ExpressionEngine/Addons/channel/mod.channel.php]: JavaScript good practices invading PHP code!

File: system/ee/ExpressionEngine/Addons/channel/mod.channel.php (hunk 3 above)
Comment (intoeetive) [reply system/ee/ExpressionEngine/Addons/channel/mod.channel.php]: It is the other way for classes and functions. Easy to get confused!

File: system/ee/ExpressionEngine/Addons/channel/mod.channel.php [hunk 6]
```diff
@@ -3985,8 +3985,18 @@ public function category_heading()
...
+                (
+                    $qstring !== ''
+                    && $this->reserved_cat_segment !== ''
+                    && in_array($this->reserved_cat_segment, explode('/', $qstring))
+                )
+                OR ee()->TMPL->fetch_param('category_url_title')
+            )
+            && ee()->TMPL->fetch_param('channel')
+        )
+        {
```
Comment (intoeetive) [inline system/ee/ExpressionEngine/Addons/channel/mod.channel.php]: Just found this line that also does not follow PSR

**Review by intoeetive — DISMISSED** [review]
Looks good now!

The tests are passing, the code looks good to me. We'll have this reviewed as well by someone else on the team, but generally I think we're good for releasing it (likely in EE 6.2)

Thank you very much for the contribution!

**Review by intoeetive — APPROVED** [review]

**Review by obfuscode — APPROVED** [review]

**Pull request merged**